base/src/numbers.rs

```diff
@@ -61,71 +74,99 @@ pub struct ExtendedU64_3(pub u64);
-    /// and <https://github.com/MinaProtocol/mina/blob/ec00ece4606244e842bf90d989d6f9bb66ab275f/src/lib/currency/currency.ml#L68>
-    pub fn to_formatted_string(&self) -> String {
-        const PRECISION: u32 = 9;
-        const PRECISION_EXP: u64 = 10_u64.pow(PRECISION);
-        let (q, r) = self.0.div_rem(&PRECISION_EXP);
-        format!("{}.{}", q, Self::pad_to_width(r, PRECISION))
+impl fmt::Display for Amount {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let (q, r) = self.0.div_rem(&MINA_PRECISION);
+        write!(f, "{}.{:#09}", q, r)
     }
+}
 
-    fn pad_to_width(r: u64, width: u32) -> String {
-        let mut s = r.to_string();
-        let num_zeros_to_pad = width - s.len() as u32;
-        for _i in 0..num_zeros_to_pad {
-            s.insert(0, '0');
+#[derive(Debug, Error, PartialEq)]
+/// Error that can be returned when parsing an Amount from string
+pub enum ParseAmountError {
+    /// Error occurs when parsing the integer components
+    #[error("Error parsing integer in Amount")]
+    ErrorParsingInteger(#[from] std::num::ParseIntError),
+
+    /// Unable to split the string on a '.' into to integer parts
+    #[error("Unexpected formatting, does not contain two integers seperated by a '.'. Got: {0}")]
+    ErrorInvalidFormat(String),
+}
+
+impl std::str::FromStr for Amount {
+    type Err = ParseAmountError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut iter = s.split('.');
+        let q: u64 = iter
+            .next()
+            .ok_or_else(|| Self::Err::ErrorInvalidFormat(s.to_string()))?
+            .parse()?;
+        let r: u64 = iter
+            .next()
+            .ok_or_else(|| Self::Err::ErrorInvalidFormat(s.to_string()))?
+            .parse()?;
+        if iter.next().is_none() {
+            // ensure there isn't more to parse as that is undefined
+            Ok(Amount(r + MINA_PRECISION * q))
+        } else {
+            Err(Self::Err::ErrorInvalidFormat(s.to_string()))
         }
-        s
     }
 }
 
-// TODO: Impl From<String> for Amount {}
-
 #[derive(Clone, Serialize, Deserialize, PartialEq, Debug, Hash, Default, WireType, From)]
 #[serde(from = "<Self as WireType>::WireType")]
 #[serde(into = "<Self as WireType>::WireType")]
+/// 4 bytes wrapped by a version
+/// Will not form part of the public API when deserialization refactor is complete
 pub struct Hex64(i64);
 
 #[derive(Clone, Serialize, Deserialize, PartialEq, Debug, Hash, Default, WireType)]
 #[serde(from = "<Self as WireType>::WireType")]
 #[serde(into = "<Self as WireType>::WireType")]
+/// A single char defined by a single byte (not variable length like a Rust char)
 pub struct Char(pub u8);
 
 #[derive(Clone, Serialize, Deserialize, PartialEq, Debug, Hash, Default, Deref, WireType, From)]
 #[serde(from = "<Self as WireType>::WireType")]
 #[serde(into = "<Self as WireType>::WireType")]
 #[wire_type(recurse = 2)]
+/// A global slot number
 pub struct GlobalSlotNumber(pub u32);
```

Comment:
unrelated, but I think `Number` is implied by the wrapped type. Also i think we already have `GlobalSlot` defined somewhere, where the field `slot_number`, can just be a `Length`, instead of `GlobalSlotNumber`.

Reply:
I'll address these types of things in the serialization refactor